### PR TITLE
PR #1001: Postgres Connection Pooling Control

### LINKED ISSUE #1001: Feature: Postgres Connection Pooling Control
**Is your feature request related to a problem? Please describe.**

I'm using a managed postgres database on Digitalocean which has pro rata connection limits via connection pooling - When running Boundary I don't see an option for controlling the number of connections made. A Boundary replica sometimes fails to start due to connection pool exhaustion. There are 3 controllers running each appearing to make 3 connections taking 9 connections total

**Describe the solution you'd like**

It would be really good to be able to specify the number of max connections a given controller can make to Postgres so I can control my usage and prevent Boundary from exhausting available connections. This could be added to Postgres configuration like below or via an env var:

```
controller {
...
  database {
    url = "postgresql://<username>:<password>@10.0.0.1:5432/<database_name>"
    max_connections = 1
  }
}

```

**Describe alternatives you've considered**

* I have considered scaling up my DO postgres database to increase possible connections which has a cost implication (This is only a pet project) 
* Scaling down my replicas
* Running a single controller pod and multiple workers

**Explain any additional use-cases**

I'm currently running Boundary on Kubernetes deployment with 3 replicas. Each replica runs as a controller and worker

**Additional context**
* My connection pool is a session pool - I had issues with transactional pools
* I am running a $15 DO Postgres instance
* Link to DO Postgres docs [https://www.digitalocean.com/docs/databases/postgresql/how-to/manage-connection-pools/](https://www.digitalocean.com/docs/databases/postgresql/how-to/manage-connection-pools/)

### PR DESCRIPTION
**Pull request by dylanrhysscott**
This closes #1001 

It appears Boundary requires at least 2 connections to start correctly presumably a connection for a controller and one for the worker? If this was set to 1 Boundary just appears to hang and server never starts so I have added a check to throw an error if 1 is specified. Setting this value to 0 will allow Boundary to open as many connections as required. Testing this using current Boundary release we see 3 idle connections from one controller:

```
 process_id | username | database_name | client_address | application_name |         backend_start         | state  |         state_change
------------+----------+---------------+----------------+------------------+-------------------------------+--------+-------------------------------
         30 |          |               |                |                  | 2021-03-17 23:13:53.71606+00  |        |
         32 | postgres |               |                |                  | 2021-03-17 23:13:53.716876+00 |        |
        167 | postgres | postgres      |                | psql             | 2021-03-17 23:47:49.24453+00  | active | 2021-03-17 23:48:59.766412+00
        173 | postgres | postgres      | 172.17.0.1     |                  | 2021-03-17 23:48:55.642789+00 | idle   | 2021-03-17 23:48:55.652682+00
        174 | postgres | postgres      | 172.17.0.1     |                  | 2021-03-17 23:48:55.658425+00 | idle   | 2021-03-17 23:48:55.700322+00
        175 | postgres | postgres      | 172.17.0.1     |                  | 2021-03-17 23:48:55.670138+00 | idle   | 2021-03-17 23:48:58.000375+00
         28 |          |               |                |                  | 2021-03-17 23:13:53.715683+00 |        |
         27 |          |               |                |                  | 2021-03-17 23:13:53.71553+00  |        |
         29 |          |               |                |                  | 2021-03-17 23:13:53.715879+00 |        |
```

Setting the following config

```
controller {
    name = "boundary-controller"
    description = "A boundary controller"
    database {
       ....
        max_open_connections = 2
    }
}
```

The number of active connections is reduced

```
 process_id | username | database_name | client_address | application_name |         backend_start         | state  |         state_change
------------+----------+---------------+----------------+------------------+-------------------------------+--------+-------------------------------
         30 |          |               |                |                  | 2021-03-17 23:13:53.71606+00  |        |
         32 | postgres |               |                |                  | 2021-03-17 23:13:53.716876+00 |        |
        167 | postgres | postgres      |                | psql             | 2021-03-17 23:47:49.24453+00  | active | 2021-03-17 23:51:08.237552+00
        180 | postgres | postgres      | 172.17.0.1     |                  | 2021-03-17 23:51:01.977899+00 | idle   | 2021-03-17 23:51:01.987556+00
        181 | postgres | postgres      | 172.17.0.1     |                  | 2021-03-17 23:51:01.991211+00 | idle   | 2021-03-17 23:51:06.271325+00
         28 |          |               |                |                  | 2021-03-17 23:13:53.715683+00 |        |
         27 |          |               |                |                  | 2021-03-17 23:13:53.71553+00  |        |
         29 |          |               |                |                  | 2021-03-17 23:13:53.715879+00 |        |
```

Setting the following config

```
controller {
    name = "boundary-controller"
    description = "A boundary controller"
    database {
       ....
        max_open_connections = 1
    }
}
```

Boundary correctly throws an error

```
➜  bin git:(postgres-conn-pool) ./boundary server -config config.hcl
Error connecting to database: unable to create db object with dialect postgres: max_open_connections must be unlimited by setting 0 or at least 2
➜  bin git:(postgres-conn-pool)
```

### CODE REVIEW

**Comment by jefferai** [comment]
I don't think this is a bad addition to have but I'd be interested in knowing why two connections are required.

**Comment by jefferai** [comment]
To put it another way, an even better solution would be to accept this patch but not error on 1 connection :-)

**Comment by dylanrhysscott** [comment]
Thanks @jefferai going to take another look at this and see if I can get some logs out of the DB driver for an explanation - Would be good to be able to minimise the number of connections for limited environments such as the ones provided by a managed Postgres option

**Comment by dylanrhysscott** [comment]
@jefferai Okay so I think I have traced the issue by littering logs everywhere for debugging - https://github.com/hashicorp/boundary/blob/main/internal/kms/repository_root_key.go#L141 this function call is not returning - I suspect it might have something to do with the reader used in list() function

**Comment by dylanrhysscott** [comment]
@jefferai Any suggestions on how to proceed with this? Wondering if its a GORM bug more so than Boundary. Is it okay to set minimum required to 2?

**Comment by jefferai** [comment]
I'm milestoning this for 0.2.1 -- you may well be right in your assessment but we want to also dig into it and verify it, regarding the reason 1 connection doesn't work. Don't worry -- we want this feature! We just haven't been able to get to it in the runup to 0.2.0...sorry about that.

**Comment by dylanrhysscott** [comment]
Hi @jefferai Thanks for getting back to me :) No worries I'm curious as well as to why one connection does not work. In theory it should I do certainly suspect this will be a GORM limitation more than a Boundary one. We may just have to set the min connections to 2 and open an issue upstream for GORM - Let me know if you need any more tests in the meantime!

File: website/content/docs/configuration/controller.mdx [hunk 1]
```diff
@@ -29,6 +30,9 @@ controller {
   - `url` - Configures the URL for connecting to Postgres
   - `migration_url` - Can be used to specify a different URL for migrations, as that
     usually requires higher privileges.
+  - `max_open_connections` - Can be used to control the maximum number of connections that can be opened by the controller.
```
Comment (jimlambrt) [inline website/content/docs/configuration/controller.mdx:33]: Perhaps update this to read:

 - `max_open_connections` - Can be used to control the maximum number of
  connections that can be opened by the controller. The minimum number of
  connections required is 2. Every controller requires 1 non-pooled
  shared lock connection (which is held for the lifetime of the controller) and
  a minimum of 1 pooled connection which is used for all other database
  requests. Setting this value to 0 will allow the controller to open as many
  connections as  needed.

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Can you update the docs, now that we have a reason for the required two connections?

File: website/content/docs/configuration/controller.mdx (hunk 1 above)
Comment (dylanrhysscott) [reply website/content/docs/configuration/controller.mdx]: Thanks for the feedback @jimlambrt have updated the docs with the specified description

**Comment by dylanrhysscott** [comment]
Changes made and re sync'd with main due to failing CI script

**Review by jimlambrt — APPROVED** [review]

**Comment by jimlambrt** [comment]
@dylanrhysscott -> ty for the contribution!

**Pull request merged**